crates/sui-json-rpc/src/coin_api.rs

```diff
@@ -288,41 +209,706 @@ impl CoinReadApiServer for CoinReadApi {
+        Ok(self.get_object_read(&object_id)?.into_object()?)
+    }
+
+    fn get_owned_coins(
+        &self,
+        owner: SuiAddress,
+        cursor: (String, ObjectID),
+        limit: usize,
+        one_coin_type_only: bool,
+    ) -> SuiResult<Vec<SuiCoin>> {
+        Ok(self
+            .state
+            .get_owned_coins_iterator_with_cursor(owner, cursor, limit, one_coin_type_only)?
+            .map(|(coin_type, coin_object_id, coin)| SuiCoin {
+                coin_type,
+                coin_object_id,
+                version: coin.version,
+                digest: coin.digest,
+                balance: coin.balance,
+                previous_transaction: coin.previous_transaction,
+            })
+            .collect::<Vec<_>>())
+    }
+
+    async fn get_executed_transaction_and_effects(
+        &self,
+        digest: TransactionDigest,
+    ) -> SuiResult<(VerifiedTransaction, TransactionEffects)> {
+        self.state
+            .get_executed_transaction_and_effects(digest)
+            .await
+    }
+
+    async fn get_balance(&self, owner: SuiAddress, coin_type: TypeTag) -> SuiResult<TotalBalance> {
+        self.state
+            .indexes
+            .as_ref()
+            .ok_or(SuiError::IndexStoreNotAvailable)?
+            .get_balance(owner, coin_type)
+            .await
+    }
+
+    async fn get_all_balance(
+        &self,
+        owner: SuiAddress,
+    ) -> SuiResult<Arc<HashMap<TypeTag, TotalBalance>>> {
+        self.state
+            .indexes
+            .as_ref()
+            .ok_or(SuiError::IndexStoreNotAvailable)?
+            .get_all_balance(owner)
+            .await
+    }
+}
+
+#[cached(
+    type = "SizedCache<String, ObjectID>",
+    create = "{ SizedCache::with_size(10000) }",
+    convert = r#"{ format!("{}{}", package_id, object_struct_tag) }"#,
+    result = true
+)]
+async fn find_package_object_id(
+    state: Arc<AuthorityState>,
+    package_id: ObjectID,
+    object_struct_tag: StructTag,
+) -> SuiApiResult<ObjectID> {
+    spawn_monitored_task!(async move {
+        let publish_txn_digest = state.find_publish_txn_digest(package_id)?;
+
+        let (_, effect) = state
+            .get_executed_transaction_and_effects(publish_txn_digest)
+            .await?;
+
+        for ((id, _, _), _) in effect.created() {
+            if let Ok(object_read) = state.get_object_read(id) {
+                if let Ok(object) = object_read.into_object() {
+                    if matches!(object.type_(), Some(type_) if type_.is(&object_struct_tag)) {
+                        return Ok(*id);
+                    }
+                }
+            }
+        }
+        Err(SuiRpcInputError::GenericNotFound(format!(
+            "Cannot find object [{}] from [{}] package event.",
+            object_struct_tag, package_id,
+        ))
+        .into())
+    })
+    .await?
+}
+
+#[cfg_attr(test, automock)]
+#[async_trait]
+pub trait CoinReadInternalTrait {
```

Comment:
nit: Do we need the `Trait` in the name here? Rust always distinguishes traits syntactically (i.e. it's impossible to confuse a trait for a type or vice versa because of context cues in the source code).